File: dart/test/envdoctor_test.dart
/// Tests for the Dart envdoctor port, adapted from the Python port's
/// `tests/test_scanner.py` and the TypeScript reference's detector tests.
library;

import 'dart:convert';
import 'dart:io';

import 'package:envdoctor/src/cli.dart' as cli;
import 'package:envdoctor/src/config/config.dart';
import 'package:envdoctor/src/core/audit.dart';
import 'package:envdoctor/src/core/model_assembler.dart';
import 'package:envdoctor/src/core/pipeline.dart';
import 'package:envdoctor/src/detectors/detector.dart';
import 'package:envdoctor/src/formatters/scan_json_formatter.dart';
import 'package:envdoctor/src/generators/env_example.dart';
import 'package:envdoctor/src/generators/environment_doc.dart';
import 'package:envdoctor/src/models/finding.dart';
import 'package:envdoctor/src/models/environment_file.dart';
import 'package:envdoctor/src/models/variable_type.dart';
import 'package:envdoctor/src/parsers/github_actions_parser.dart';
import 'package:envdoctor/src/parsers/k8s_parser.dart';
import 'package:envdoctor/src/parsers/source_parser.dart';
import 'package:envdoctor/src/utils/type_infer.dart';
import 'package:test/test.dart';

Directory _tmp() => Directory.systemTemp.createTempSync('envdoctor-dart-test-');

String _write(Directory root, String name, String content) {
  final file = File('${root.path}${Platform.pathSeparator}$name');
  file.parent.createSync(recursive: true);
  file.writeAsStringSync(content);
  return file.path;
}

/// Scan a temp project through the real pipeline and return findings.
List<Finding> _scan(Directory root) {
  final ctx = loadProject(root.path);
  return runAudit(ctx.model, AuditOptions()).findings;
}

Set<String> _vars(List<Finding> findings, String rule) =>
    findings.where((f) => f.ruleId == rule).map((f) => f.variable).toSet();

void main() {
  group('detectors', () {
    test('missing and unused reconcile against source usage', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=postgres://x\nUNUSED_KEY=1\n');
      _write(root, 'src/index.ts',
          "const a = process.env.DB_URL;\nconst b = process.env.NEW_FLAG;\n");

      final findings = _scan(root);
      final undefined = _vars(findings, 'undefined-in-source');
      final unused = _vars(findings, 'unused');

      expect(undefined, contains('NEW_FLAG')); // used but not defined
      expect(unused, contains('UNUSED_KEY')); // defined but not used
      expect(undefined, isNot(contains('DB_URL'))); // reconciled
      expect(unused, isNot(contains('DB_URL')));
      // Values must never leak into findings.
      for (final f in findings) {
        expect(f.message, isNot(contains('postgres')));
      }
    });

    test('duplicates flags a key defined twice in one file', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=a\nDB_URL=b\nSOLO=1\n');
      _write(root, 'src/index.ts',
          'process.env.DB_URL;\nprocess.env.SOLO;\n');

      final findings = _scan(root);
      final dups = findings.where((f) => f.ruleId == 'duplicates').toList();
      expect(dups.map((f) => f.variable), contains('DB_URL'));
      expect(dups.map((f) => f.variable), isNot(contains('SOLO')));
      expect(dups.first.severity, Severity.error);
      // The first occurrence still reconciles: not reported unused.
      expect(_vars(findings, 'unused'), isNot(contains('DB_URL')));
    });

    test('public-prefix flags secret-like names behind public prefixes', () {
      final root = _tmp();
      _write(root, '.env',
          'NEXT_PUBLIC_API_KEY=x\nPUBLIC_URL=x\nAPI_KEY=x\n');

      final findings = _scan(root);
      final names = _vars(findings, 'public-prefix');
      expect(names, {'NEXT_PUBLIC_API_KEY'});
      expect(
        findings.firstWhere((f) => f.ruleId == 'public-prefix').severity,
        Severity.error,
      );
    });

    test('weak-secret flags placeholder and short values without leaking', () {
      final root = _tmp();
      _write(root, '.env',
          'API_KEY=changeme\nSTRONG_TOKEN=s0m3-l0ng-r4nd0m-value\nSHORT_SECRET=abc\n');

      final findings = _scan(root);
      final weak = _vars(findings, 'weak-secret');
      expect(weak, contains('API_KEY')); // placeholder
      expect(weak, contains('SHORT_SECRET')); // too short
      expect(weak, isNot(contains('STRONG_TOKEN')));
      for (final f in findings) {
        expect(f.message, isNot(contains('changeme')));
        expect(f.message, isNot(contains('s0m3')));
      }
    });

    test('typo suggests the closest defined name', () {
      final root = _tmp();
      _write(root, '.env', 'DATABASE_URL=postgres://x\n');
      _write(root, 'src/index.ts', 'process.env.DATBASE_URL;\n');

      final findings = _scan(root);
      final typos = findings.where((f) => f.ruleId == 'typo').toList();
      expect(typos, hasLength(1));
      expect(typos.single.variable, 'DATBASE_URL');
      expect(typos.single.severity, Severity.warning);
      expect(typos.single.message, contains('did you mean "DATABASE_URL"'));
    });

    test('environment-diff reports variables missing from one environment', () {
      final root = _tmp();
      _write(root, '.env', 'A=1\nB=2\nC=3\n');
      _write(root, '.env.production', 'A=1\nC=3\nD=4\n');

      final names = _vars(_scan(root), 'environment-diff');
      expect(names, containsAll(['B', 'D']));
      expect(names, isNot(contains('A')));
      expect(names, isNot(contains('C')));
    });

    test('type-mismatch flags incompatible types across environments', () {
      final root = _tmp();
      _write(root, '.env', 'PORT=3000\nHOST=8080\n');
      _write(root, '.env.production', 'PORT=abc\nHOST=9090\n');

      final findings = _scan(root);
      final mismatches = _vars(findings, 'type-mismatch');
      expect(mismatches, contains('PORT')); // integer vs string
      expect(mismatches, isNot(contains('HOST'))); // integer vs integer
      final tm = findings.firstWhere((f) => f.ruleId == 'type-mismatch');
      expect(tm.severity, Severity.error);
      expect(tm.message, contains('expected: integer'));
      expect(tm.message, isNot(contains('3000')));
      expect(tm.message, isNot(contains('abc')));
    });

    test('infra sources (compose, actions, k8s) are scanned', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=postgres://x\n');
      _write(
        root,
        'docker-compose.yml',
        'services:\n'
            '  web:\n'
            '    image: x\n'
            '    environment:\n'
            '      - TOKEN=\${COMPOSE_SECRET}\n'
            '      - DB=\${DB_URL}\n',
      );
      _write(
        root,
        '.github/workflows/ci.yml',
        'jobs:\n'
            '  build:\n'
            '    steps:\n'
            '      - run: deploy\n'
            '        env:\n'
            '          KEY: \${{ secrets.DEPLOY_KEY }}\n'
            '          REGION: \${{ vars.REGION }}\n',
      );
      _write(
        root,
        'k8s/app.deployment.yaml',
        'apiVersion: apps/v1\n'
            'kind: Deployment\n'
            'spec:\n'
            '  template:\n'
            '    spec:\n'
            '      containers:\n'
            '        - name: app\n'
            '          command: ["sh", "-c", "echo \${K8S_VAR}"]\n'
            '          env:\n'
            '            - name: PORT\n'
            '              value: "3000"\n',
      );

      final findings = _scan(root);
      final referenced =
          _vars(findings, 'undefined-in-source').union(_vars(findings, 'missing'));
      // Referenced only in compose but never defined anywhere.
      // (The reference flags only compose references as `missing`; k8s and
      // GitHub Actions usages feed the other detectors instead.)
      expect(referenced, containsAll(['COMPOSE_SECRET', 'TOKEN', 'DB']));
      // GitHub Actions env keys, secrets, and vars are external: never flagged.
      final flagged = findings.map((f) => f.variable).toSet();
      expect(flagged.intersection({'DEPLOY_KEY', 'REGION', 'KEY'}), isEmpty);
      // DB_URL is defined and referenced in compose -> NOT unused.
      expect(_vars(findings, 'unused'), isNot(contains('DB_URL')));
      // No values leak into any finding.
      for (final f in findings) {
        expect(f.message, isNot(contains('postgres')));
      }
    });

    test('schema-validation flags values violating the configured schema', () {
      final root = _tmp();
      final envPath = _write(root, '.env', 'PORT=80\nNODE_ENV=dev\n');
      final config = EnvdoctorConfig()
        ..schema = {
          'PORT': VariableSchema(type: SchemaType.integer, min: 1024),
          'NODE_ENV': VariableSchema(enumValues: [
            'development',
            'production',
            'test',
          ]),
        };
      final model = assembleModel(root.path, config, [envPath]);
      final detector =
          allDetectors().firstWhere((d) => d.id == 'schema-validation');
      final findings = detector.detect(IndexedModel.buildIndex(model));
      expect(findings.map((f) => f.variable).toSet(), {'PORT', 'NODE_ENV'});
      // The offending values must never appear in findings.
      final portMsg = findings.firstWhere((f) => f.variable == 'PORT').message;
      expect(portMsg, isNot(contains('80')));
    });
  });

  group('parsers', () {
    test('k8s parser extracts container env defs and command interpolations', () {
      final parser = K8sParser();
      final file = parser.parse(
        'apiVersion: apps/v1\n'
        'kind: Deployment\n'
        'spec:\n'
        '  template:\n'
        '    spec:\n'
        '      containers:\n'
        '        - name: app\n'
        '          command: ["sh", "-c", "echo \${MESSAGE}"]\n'
        '          env:\n'
        '            - name: PORT\n'
        '              value: "3000"\n'
        '            - name: SECRET_TOKEN\n'
        '              valueFrom:\n'
        '                secretKeyRef:\n'
        '                  name: app-secret\n'
        '                  key: token\n',
        '/p/k8s/deployment.yaml',
      );
      final defined = file.variables.map((v) => v.name).toSet();
      expect(defined, containsAll(['PORT']));
      expect(defined, isNot(contains('SECRET_TOKEN')));
      final used = file.usages.map((v) => v.name).toSet();
      expect(used, containsAll(['MESSAGE', 'SECRET_TOKEN']));
    });

    test('actions parser extracts env keys and secrets/vars references', () {
      final parser = GithubActionsParser();
      final file = parser.parse(
        'name: ci\n'
        'on: [push]\n'
        'env:\n'
        "  CI: 'true'\n"
        'jobs:\n'
        '  t:\n'
        '    runs-on: ubuntu-latest\n'
        '    steps:\n'
        '      - run: echo \${{ secrets.DEPLOY_TOKEN }}\n'
        '      - run: echo \${{ vars.REGION }}\n'
        '        env:\n'
        '          KEY: \${{ env.BUILD_ID }}\n',
        '/p/.github/workflows/ci.yml',
      );
      // Nothing here is expected in .env files: no missing findings may
      // result from these. The parser still records what it saw.
      expect(file.format, FileFormat.githubActions);
    });

    test('source parser finds every supported access pattern', () {
      final parser = SourceParser(['ts', 'js']);
      final file = parser.parse(
        'const a = process.env.PLAIN;\n'
        'const b = process.env["BRACKET"];\n'
        "const c = process.env['SINGLE_BRACKET'];\n"
        'const d = import.meta.env.VITE_KEY;\n'
        '// const e = process.env.COMMENTED_OUT;\n'
        'const f = "process.env.IN_STRING";\n',
        '/p/src/index.ts',
      );
      final names = file.usages.map((v) => v.name).toSet();
      expect(names,
          containsAll(['PLAIN', 'BRACKET', 'SINGLE_BRACKET', 'VITE_KEY']));
      expect(names, isNot(contains('COMMENTED_OUT')));
      expect(names, isNot(contains('IN_STRING')));
    });
  });

  group('type inference', () {
    test('infers scalar types', () {
      expect(inferType(null), VariableType.unknown);
      expect(inferType(''), VariableType.unknown);
      expect(inferType('3000'), VariableType.integer);
      expect(inferType('-7'), VariableType.integer);
      expect(inferType('1.5'), VariableType.float);
      expect(inferType('TRUE'), VariableType.boolean);
      expect(inferType('https://x.io'), VariableType.url);
      expect(inferType('{"a":1}'), VariableType.json);
      expect(inferType('hello'), VariableType.string);
    });
  });

  group('generators', () {
    test('env example and docs never contain secret values', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=postgres://x\nAPI_KEY=topsecretvalue\n');
      _write(root, 'src/index.ts', 'process.env.PORT;\n');

      final ctx = loadProject(root.path);
      final example = generateEnvExample(ctx.model, ctx.config);
      final doc = generateEnvironmentDoc(ctx.model, ctx.config);

      expect(example, contains('DB_URL'));
      expect(example, contains('PORT'));
      expect(doc, contains('DB_URL'));
      // Secret-like names always get a blank value.
      expect(example, contains('API_KEY=\n'));
      expect(example, isNot(contains('topsecretvalue')));
      expect(doc, isNot(contains('topsecretvalue')));
    });
  });

  group('scan JSON output', () {
    test('matches the reference shape and leaks no values', () {
      final root = _tmp();
      _write(root, '.env', 'API_KEY=changeme\n');
      _write(root, 'src/index.ts', 'process.env.DATBASE_URL;\n');

      final ctx = loadProject(root.path);
      final result = runAudit(ctx.model, AuditOptions());
      final out =
          renderScanJson(root.path, result.findings, result.summary, result.exitCode);
      final data = jsonDecode(out) as Map<String, dynamic>;

      expect(data.keys, containsAll(['exitCode', 'summary', 'findings']));
      final summary = data['summary'] as Map<String, dynamic>;
      expect(
        summary.keys,
        containsAll([
          'filesScanned',
          'variablesFound',
          'errors',
          'warnings',
          'infos',
          'total',
        ]),
      );
      final findings = data['findings'] as List<dynamic>;
      expect(findings, isNotEmpty);
      for (final obj in findings.cast<Map<String, dynamic>>()) {
        expect(
          obj.keys,
          containsAll(['id', 'ruleId', 'severity', 'variable', 'message', 'locations']),
        );
      }
      expect(out, isNot(contains('changeme')));
    });
  });

  group('CLI', () {
    test('scan exits non-zero when errors are found', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=x\n');
      _write(root, 'src/index.ts', 'process.env.MISSING_VAR;\n');

      expect(cli.run(['scan', '--json', '-C', root.path]), isNot(0));
    });

    test('scan exits zero on a clean project', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=x\n');
      _write(root, 'src/index.ts', 'process.env.DB_URL;\n');

      expect(cli.run(['scan', '-C', root.path]), 0);
    });

    test('--version and help work', () {
      expect(cli.run(['--version']), 0);
      expect(cli.run(['--help']), 0);
    });
  });

  group('commands', () {
    test('init creates, skips, and --force rewrites', () {
      final root = _tmp();
      _write(root, '.env', 'DB_URL=postgres://x\nAPI_KEY=topsecretvalue\n');
      _write(root, 'src/index.ts', 'process.env.PORT;\n');

      expect(cli.run(['init', '-C', root.path]), 0);
      final example = File('${root.path}/.env.example');
      final envdoc = File('${root.path}/ENVIRONMENT.md');
      expect(example.existsSync(), isTrue);
      expect(envdoc.existsSync(), isTrue);
      final firstExample = example.readAsStringSync();
      // Secret values are never written into generated artifacts.
      expect(firstExample, isNot(contains('topsecretvalue')));
      expect(envdoc.readAsStringSync(), isNot(contains('topsecretvalue')));

      // Second init without --force skips; files unchanged.
      example.writeAsStringSync('SENTINEL');
      expect(cli.run(['init', '-C', root.path]), 0);
      expect(example.readAsStringSync(), 'SENTINEL');

      // --force rewrites.
      expect(cli.run(['init', '-C', root.path, '--force']), 0);
      expect(example.readAsStringSync(), firstExample);
    });

    test('diff and sync copy keys without values', () {
      final root = _tmp();
      _write(root, '.env', 'A=1\nB=2\n');
      _write(root, '.env.production', 'A=9\n');

      // diff exits 1 when variables are missing on one side (TS parity).
      expect(cli.run(['diff', 'development', 'production', '-C', root.path]), 1);
      expect(cli.run(['sync', 'development', 'production', '-C', root.path]), 0);

      final prod =
          File('${root.path}/.env.production').readAsStringSync();
      expect(prod, contains('B='));
      expect(prod, contains('A=9'));
      expect(prod, isNot(contains('B=2'))); // value never copied
    });
  });
}

File: elixir/test/envdoctor_test.exs
defmodule EnvdoctorTest do
  use ExUnit.Case, async: false

  import ExUnit.CaptureIO

  alias Envdoctor.{CLI, Scanner}
  alias Envdoctor.Runtime.{Compare, Token}
  alias Envdoctor.Scanner.ScanResult

  setup do
    dir = Path.join(System.tmp_dir!(), "envdoctor_test_#{:erlang.unique_integer([:positive])}")
    File.mkdir_p!(dir)
    on_exit(fn -> File.rm_rf!(dir) end)
    {:ok, dir: dir}
  end

  defp write(dir, name, content) do
    path = Path.join(dir, name)
    File.mkdir_p!(Path.dirname(path))
    File.write!(path, content)
    path
  end

  defp errors(%ScanResult{} = r), do: ScanResult.errors(r)
  defp warnings(%ScanResult{} = r), do: ScanResult.warnings(r)

  test "version matches the reference" do
    assert Envdoctor.version() == "0.1.2"
  end

  test "detects all usage forms and ignores comments/heredocs", %{dir: dir} do
    src =
      write(dir, "app.ex", """
      defmodule App do
        # System.get_env("COMMENTED")
        @moduledoc \"\"\"
        Example: System.get_env("DOC_IGNORED")
        \"\"\"
        def a, do: System.get_env("DB_URL")
        def b, do: System.get_env("PORT", "3000")
        def c, do: System.fetch_env("API_KEY")
        def d, do: System.fetch_env!("HOST")
      end
      """)

    used = Scanner.scan_source_file(src)
    assert Map.keys(used) |> Enum.sort() == ["API_KEY", "DB_URL", "HOST", "PORT"]
    refute Map.has_key?(used, "COMMENTED")
    refute Map.has_key?(used, "DOC_IGNORED")
  end

  test "detects JS usage forms when js extensions are scanned", %{dir: dir} do
    src =
      write(dir, "app.js", """
      // process.env.COMMENTED
      /* process.env.BLOCKED */
      const a = process.env.DATABASE_URL;
      const b = process.env["API_KEY"];
      const c = import.meta.env.VITE_FLAG;
      """)

    used = Scanner.scan_source_file(src)
    assert Map.keys(used) |> Enum.sort() == ["API_KEY", "DATABASE_URL", "VITE_FLAG"]
    refute Map.has_key?(used, "COMMENTED")
    refute Map.has_key?(used, "BLOCKED")
  end

  test "missing and unused", %{dir: dir} do
    write(dir, ".env", "DB_URL=postgres://x\nUNUSED_KEY=1\n")
    write(dir, "app.ex", "System.get_env(\"DB_URL\")\nSystem.get_env(\"NEW_FLAG\")\n")

    result = Scanner.scan(dir)
    err_names = result |> errors() |> Enum.map(& &1.name)
    warn_names = result |> warnings() |> Enum.map(& &1.name)

    assert "NEW_FLAG" in err_names
    assert "UNUSED_KEY" in warn_names
    refute "DB_URL" in err_names
    refute "DB_URL" in warn_names
  end

  test "duplicates", %{dir: dir} do
    write(dir, ".env", "DB_URL=a\nDB_URL=b\nSOLO=1\n")
    write(dir, "app.ex", "System.get_env(\"DB_URL\")\nSystem.get_env(\"SOLO\")\n")

    result = Scanner.scan(dir)
    dups = Enum.filter(result.findings, &(&1.rule == "duplicates"))
    assert length(dups) == 1
    assert hd(dups).name == "DB_URL"
    assert hd(dups).severity == "error"
    assert hd(dups).message =~ "lines 1, 2"
    refute Enum.any?(result.findings, &(&1.rule == "unused" and &1.name == "DB_URL"))
  end

  test "public-prefix", %{dir: dir} do
    write(dir, ".env", "NEXT_PUBLIC_API_KEY=x\nPUBLIC_URL=x\nAPI_KEY=x\n")
    result = Scanner.scan(dir)
    pp = Enum.filter(result.findings, &(&1.rule == "public-prefix"))
    assert Enum.map(pp, & &1.name) == ["NEXT_PUBLIC_API_KEY"]
    assert hd(pp).severity == "error"
  end

  test "clean project has no findings", %{dir: dir} do
    write(dir, ".env", "DB_URL=x\n")
    write(dir, "app.ex", "System.get_env(\"DB_URL\")\n")
    assert Scanner.scan(dir).findings == []
  end

  test "infer_type" do
    assert Scanner.infer_type("") == "empty"
    assert Scanner.infer_type("3000") == "integer"
    assert Scanner.infer_type("-7") == "integer"
    assert Scanner.infer_type("1.5") == "float"
    assert Scanner.infer_type("TRUE") == "boolean"
    assert Scanner.infer_type("https://x.io") == "url"
    assert Scanner.infer_type(~s({"a":1})) == "json"
    assert Scanner.infer_type("hello") == "string"
  end

  test "levenshtein" do
    assert Scanner.levenshtein("abc", "abc") == 0
    assert Scanner.levenshtein("DATBASE_URL", "DATABASE_URL") == 1
    assert Scanner.levenshtein("", "abc") == 3
  end

  test "weak-secret never leaks values", %{dir: dir} do
    write(dir, ".env", "API_KEY=changeme\nSTRONG_TOKEN=s0m3-l0ng-r4nd0m-value\nSHORT_SECRET=abc\n")
    write(dir, "app.ex", "# nothing used\n")

    result = Scanner.scan(dir)
    weak = result.findings |> Enum.filter(&(&1.rule == "weak-secret")) |> Enum.map(& &1.name)
    assert "API_KEY" in weak
    assert "SHORT_SECRET" in weak
    refute "STRONG_TOKEN" in weak

    for f <- result.findings do
      assert f.severity in ["error", "warning"]
      refute f.message =~ "changeme"
      refute f.message =~ "s0m3"
    end
  end

  test "typo suggestion", %{dir: dir} do
    write(dir, ".env", "DATABASE_URL=postgres://x\n")
    write(dir, "app.ex", "System.get_env(\"DATBASE_URL\")\n")

    result = Scanner.scan(dir)
    typos = Enum.filter(result.findings, &(&1.rule == "typo"))
    assert length(typos) == 1
    assert hd(typos).name == "DATBASE_URL"
    assert hd(typos).severity == "warning"
    assert hd(typos).message =~ ~s(did you mean "DATABASE_URL")
    assert "DATBASE_URL" in Enum.map(errors(result), & &1.name)
  end

  test "environment-diff", %{dir: dir} do
    write(dir, ".env", "SHARED=1\n")
    write(dir, ".env.production", "SHARED=1\nONLY_PROD=1\n")
    write(dir, "app.ex", "# none\n")

    result = Scanner.scan(dir)
    diffs = Map.new(Enum.filter(result.findings, &(&1.rule == "environment-diff")), &{&1.name, &1})
    assert Map.has_key?(diffs, "ONLY_PROD")
    assert diffs["ONLY_PROD"].severity == "warning"
    assert diffs["ONLY_PROD"].message =~ "default"
    assert diffs["ONLY_PROD"].message =~ "production"
    refute Map.has_key?(diffs, "SHARED")
  end

  test "type-mismatch", %{dir: dir} do
    write(dir, ".env", "PORT=3000\nHOST=8080\n")
    write(dir, ".env.production", "PORT=abc\nHOST=9090\n")
    write(dir, "app.ex", "# none\n")

    result = Scanner.scan(dir)
    mismatches = result.findings |> Enum.filter(&(&1.rule == "type-mismatch")) |> Enum.map(& &1.name)
    assert "PORT" in mismatches
    refute "HOST" in mismatches

    tm = Enum.find(result.findings, &(&1.rule == "type-mismatch"))
    assert tm.severity == "error"
    refute tm.message =~ "3000"
    refute tm.message =~ "abc"
  end

  test "scan --json output shape and no value leak", %{dir: dir} do
    write(dir, ".env", "API_KEY=changeme\n")
    write(dir, "app.ex", "System.get_env(\"DATBASE_URL\")\n")
    write(dir, ".env.production", "API_KEY=changeme\nDATABASE_URL=x\n")

    out = capture_io(fn -> assert CLI.run(["scan", "--dir", dir, "--json"]) == 1 end)
    {:ok, data} = Envdoctor.Json.decode(out)
    assert is_list(data)

    for obj <- data do
      assert obj |> Map.keys() |> Enum.sort() == ["file", "line", "message", "name", "rule", "severity"]
    end

    refute out =~ "changeme"
  end

  test "finding_to_map shape", %{dir: dir} do
    write(dir, ".env", "UNUSED=1\n")
    write(dir, "app.ex", "# none\n")
    [finding | _] = Scanner.scan(dir).findings
    d = Scanner.finding_to_map(finding, dir)
    assert d["file"] == ".env"
    assert is_integer(d["line"])
  end

  test "infra sources scanned (compose/actions/k8s)", %{dir: dir} do
    write(dir, ".env", "DB_URL=postgres://x\n")

    write(dir, "docker-compose.yml", """
    services:
      web:
        image: x
        environment:
          - TOKEN=${COMPOSE_SECRET}
          - DB=${DB_URL}
    """)

    write(dir, ".github/workflows/ci.yml", """
    jobs:
      build:
        steps:
          - run: deploy
            env:
              KEY: ${{ secrets.DEPLOY_KEY }}
              REGION: ${{ vars.REGION }}
    """)

    write(dir, "k8s/deploy.yaml", """
    apiVersion: apps/v1
    kind: Deployment
    spec:
      value: ${K8S_VAR}
    """)

    result = Scanner.scan(dir)

    undefined =
      result |> errors() |> Enum.filter(&(&1.rule == "undefined-in-source")) |> Enum.map(& &1.name)

    for name <- ["COMPOSE_SECRET", "DEPLOY_KEY", "REGION", "K8S_VAR"] do
      assert name in undefined
    end

    for f <- result.findings, f.rule == "undefined-in-source" do
      assert f.message == "referenced but not defined in any environment file"
    end

    unused = result |> warnings() |> Enum.filter(&(&1.rule == "unused")) |> Enum.map(& &1.name)
    refute "DB_URL" in unused

    for f <- result.findings do
      refute f.message =~ "postgres"
    end
  end

  test "generate_docs exact content", %{dir: dir} do
    write(dir, ".env", "DB_URL=postgres://x\n")
    write(dir, "app.ex", "System.get_env(\"PORT\")\n")

    docs = Scanner.generate_docs(dir)

    assert docs[".env.example"] ==
             "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n"

    md = docs["ENVIRONMENT.md"]
    assert String.starts_with?(md, "# Environment variables\n\n| Variable | Defined | Used |\n")
    assert md =~ "| DB_URL | yes | no |"
    assert md =~ "| PORT | no | yes |"
    assert String.ends_with?(md, "\n")
    refute docs[".env.example"] =~ "postgres"
    refute md =~ "postgres"
  end

  test "init creates, skips, and forces", %{dir: dir} do
    write(dir, ".env", "DB_URL=postgres://x\n")
    write(dir, "app.ex", "System.get_env(\"PORT\")\n")

    example = Path.join(dir, ".env.example")
    envdoc = Path.join(dir, "ENVIRONMENT.md")

    out = capture_io(fn -> assert CLI.run(["init", "--dir", dir]) == 0 end)
    assert out =~ "created .env.example"
    assert out =~ "created ENVIRONMENT.md"

    first_example = File.read!(example)
    first_md = File.read!(envdoc)

    assert first_example ==
             "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n"

    refute first_example =~ "postgres"
    refute first_md =~ "postgres"

    File.write!(example, "SENTINEL")
    out = capture_io(fn -> assert CLI.run(["init", "--dir", dir]) == 0 end)
    assert out =~ "skipped .env.example (exists)"
    assert out =~ "skipped ENVIRONMENT.md (exists)"
    assert File.read!(example) == "SENTINEL"

    out = capture_io(fn -> assert CLI.run(["init", "--dir", dir, "--force"]) == 0 end)
    assert out =~ "wrote .env.example"
    assert out =~ "wrote ENVIRONMENT.md"
    assert File.read!(example) == first_example
  end

  test "fix always rewrites", %{dir: dir} do
    write(dir, ".env", "DB_URL=postgres://x\n")
    write(dir, "app.ex", "System.get_env(\"PORT\")\n")
    write(dir, ".env.example", "STALE")

    out = capture_io(fn -> assert CLI.run(["fix", "--dir", dir]) == 0 end)
    assert out =~ "wrote .env.example"
    assert out =~ "wrote ENVIRONMENT.md"

    assert File.read!(Path.join(dir, ".env.example")) ==
             "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n"

    md = File.read!(Path.join(dir, "ENVIRONMENT.md"))
    assert md =~ "| DB_URL | yes | no |"
    assert md =~ "| PORT | no | yes |"
    refute md =~ "postgres"
  end

  test "diff and sync", %{dir: dir} do
    write(dir, ".env", "A=1\nB=2\n")
    write(dir, ".env.production", "A=9\n")

    assert Scanner.diff_labels(dir, "default", "production") ==
             %{"onlyInA" => ["B"], "onlyInB" => [], "common" => ["A"]}

    # dry-run does not write
    assert Scanner.sync_labels(dir, "default", "production", true) == ["B"]
    refute File.read!(Path.join(dir, ".env.production")) =~ "B="

    # real sync appends B= (no value copied) and leaves A's value intact
    assert Scanner.sync_labels(dir, "default", "production") == ["B"]
    prod = File.read!(Path.join(dir, ".env.production"))
    assert prod =~ "B=\n"
    assert prod =~ "A=9"
    refute prod =~ "B=2"
    assert Scanner.diff_labels(dir, "default", "production")["common"] == ["A", "B"]
  end

  test "schema validation", %{dir: dir} do
    write(dir, ".env", "PORT=99999\nLEVEL=verbose\nAPI=ftp://x\nGOOD=info\n")

    write(dir, "envdoctor.schema.json", ~s|{"PORT":{"type":"integer","max":65535},"LEVEL":{"enum":["debug","info"]},"API":{"type":"url"},"MISSING":{"type":"string"},"GOOD":{"enum":["info","warn"]}}|)

    result = Scanner.scan(dir)

    schema =
      Map.new(
        Enum.filter(result.findings, &(&1.rule == "schema-validation")),
        &{&1.name, &1.message}
      )

    assert schema == %{
             "PORT" => "value exceeds the maximum",
             "LEVEL" => "value is not one of the allowed values",
             "API" => "value does not match schema type url",
             "MISSING" => "required by schema but not defined"
           }

    for f <- result.findings do
      refute f.message =~ "99999"
      refute f.message =~ "verbose"
    end
  end

  test "schema optional and absent", %{dir: dir} do
    write(dir, ".env", "A=1\n")
    assert Scanner.load_schema(dir) == %{}
    refute Enum.any?(Scanner.scan(dir).findings, &(&1.rule == "schema-validation"))

    write(dir, "envdoctor.schema.json", ~s|{"MAYBE":{"type":"string","optional":true}}|)
    refute Enum.any?(Scanner.scan(dir).findings, &(&1.rule == "schema-validation"))
  end

  test "scan exits 1 on errors, 0 when clean, 1 with --strict warnings", %{dir: dir} do
    write(dir, ".env", "UNUSED=1\n")
    write(dir, "app.ex", "# none\n")
    capture_io(fn -> assert CLI.run(["scan", "--dir", dir]) == 0 end)
    capture_io(fn -> assert CLI.run(["scan", "--dir", dir, "--strict"]) == 1 end)

    write(dir, "app.ex", "System.get_env(\"MISSING_VAR\")\n")
    capture_io(fn -> assert CLI.run(["scan", "--dir", dir]) == 1 end)
  end

  test "token round-trip and snapshot-diff equivalence" do
    snapshot = %{
      schema: 1,
      capturedAt: "2025-01-01T00:00:00Z",
      os: %{platform: "darwin", arch: "arm64", release: "24.0.0"},
      tools: [%{tool: "node", version: "20.11.1", resolvedFrom: "/usr/local/bin"}],
      path: ["/usr/local/bin", "/usr/bin"],
      globals: %{},
      envFlagNames: ["HOME", "PATH"]
    }

    token = Token.encode(snapshot)
    assert String.starts_with?(token, "envd1:")
    assert {:ok, decoded} = Token.decode(token)
    assert decoded.tools == [%{tool: "node", version: "20.11.1", resolvedFrom: "/usr/local/bin"}]

    diff = Compare.compare(decoded, decoded)
    assert diff.equivalent

    other = put_in(decoded.tools, [%{tool: "node", version: "22.0.0", resolvedFrom: "/opt/bin"}])
    diff = Compare.compare(decoded, other)
    refute diff.equivalent
    assert Enum.any?(diff.tools, &(&1.name == "node" and &1.status == "different"))

    assert {:error, _} = Token.decode("not-a-token")
    assert {:error, _} = Token.decode("envd1:!!!garbage!!!")
  end

  test "snapshot capture records names only, never secret names" do
    snapshot = Envdoctor.capture_snapshot()
    assert snapshot.schema == 1
    assert is_binary(snapshot.capturedAt)
    assert is_binary(snapshot.os.platform)
    assert is_list(snapshot.path)
    assert is_list(snapshot.envFlagNames)
    # PATH itself is an env name; values never appear in the snapshot.
    refute Map.has_key?(snapshot, :env)
  end
end
